esphome/components/ld24xx/ld24xx.h

```diff
@@ -0,0 +1,50 @@
+#pragma once
+
+#include <memory>
+
+#ifdef USE_SENSOR
+#include "esphome/components/sensor/sensor.h"
+
+// NOLINTBEGIN
+#define SUB_SENSOR_WITH_DEDUP(name, dedup_type) \
+ protected: \
+  ld24xx::SensorWithDedup<dedup_type> name##_sensor_; \
+\
+ public: \
+  void set_##name##_sensor(sensor::Sensor *sensor) { \
+    this->name##_sensor_.sens = sensor; \
+    this->name##_sensor_.publish_dedup = std::make_unique<Deduplicator<dedup_type>>(); \
+  }
+#endif
+// NOLINTEND
+
+namespace esphome {
+namespace ld24xx {
+
+#ifdef USE_SENSOR
+// Helper class to store a sensor with a deduplicator & publish state only when the value changes
+template<typename T> class SensorWithDedup {
+ public:
+  void publish_state_if_not_dup(T state) {
+    if (this->sens != nullptr && (this->publish_dedup->next(state) || this->state_unknown_)) {
+      this->sens->publish_state(static_cast<float>(state));
+      this->state_unknown_ = false;
+    }
+  }
+
+  void publish_state_unknown() {
+    if (this->sens != nullptr && !this->state_unknown_) {
+      this->sens->publish_state(NAN);
+      this->state_unknown_ = true;
+    }
+  }
+
+  sensor::Sensor *sens{nullptr};
+  std::unique_ptr<Deduplicator<T>> publish_dedup;
```

Comment:
```suggestion
  Deduplicator<T> publish_dedup;
```

Do we need `std::unique_ptr` here?

Reply:
```diff
diff --git a/esphome/components/ld2410/ld2410.cpp b/esphome/components/ld2410/ld2410.cpp
index 712b701ef..1f594eca7 100644
--- a/esphome/components/ld2410/ld2410.cpp
+++ b/esphome/components/ld2410/ld2410.cpp
@@ -785,12 +785,10 @@ void LD2410Component::set_light_out_control() {
 #ifdef USE_SENSOR
 void LD2410Component::set_gate_move_sensor(uint8_t gate, sensor::Sensor *s) {
   this->gate_move_sensors_[gate].sens = s;
-  this->gate_move_sensors_[gate].publish_dedup = std::make_unique<Deduplicator<uint8_t>>();
 }
 
 void LD2410Component::set_gate_still_sensor(uint8_t gate, sensor::Sensor *s) {
   this->gate_still_sensors_[gate].sens = s;
-  this->gate_still_sensors_[gate].publish_dedup = std::make_unique<Deduplicator<uint8_t>>();
 }
 #endif
 
diff --git a/esphome/components/ld24xx/ld24xx.h b/esphome/components/ld24xx/ld24xx.h
index b9c832882..60b1f0b37 100644
--- a/esphome/components/ld24xx/ld24xx.h
+++ b/esphome/components/ld24xx/ld24xx.h
@@ -1,7 +1,5 @@
 #pragma once
 
-#include <memory>
-
 #ifdef USE_SENSOR
 #include "esphome/components/sensor/sensor.h"
 
@@ -13,7 +11,6 @@
  public: \
   void set_##name##_sensor(sensor::Sensor *sensor) { \
     this->name##_sensor_.sens = sensor; \
-    this->name##_sensor_.publish_dedup = std::make_unique<Deduplicator<dedup_type>>(); \
   }
 #endif
 // NOLINTEND
@@ -26,7 +23,7 @@ namespace ld24xx {
 template<typename T> class SensorWithDedup {
  public:
   void publish_state_if_not_dup(T state) {
-    if (this->sens != nullptr && (this->publish_dedup->next(state) || this->state_unknown_)) {
+    if (this->sens != nullptr && (this->publish_dedup.next(state) || this->state_unknown_)) {
       this->sens->publish_state(static_cast<float>(state));
       this->state_unknown_ = false;
     }
@@ -40,7 +37,7 @@ template<typename T> class SensorWithDedup {
   }
 
   sensor::Sensor *sens{nullptr};
-  std::unique_ptr<Deduplicator<T>> publish_dedup;
+  Deduplicator<T> publish_dedup;
 
  protected:
   bool state_unknown_{false};
```

Reply:
Or is it that only one of them is used so you wanted to allocate only some of them ?

Reply:
So some of them will never be used if the user doesn't enable them.

Got it